### ISSUE TITLE
GSSAPI rekey re-uses existing credentials

### ISSUE THREAD

**Issue by PromyLOPh** [ronf/asyncssh#476]
We have a fairly complex setup using SSSD and Kerberos. SSSD is set to re-new issued tickets automatically using sssd-kcm to enable long-running sessions. We use asyncssh in a custom client to connect via SSH and authenticate by forwarding Kerberos tickets, i.e. `GSSAPIKeyExchange yes` and `GSSAPIAuthentication yes`. But our clients keep disconnecting after about 9 hours (8 hours is the ticket expiry time and sshd seems to rekey every hour) with a cryptic exception

```
asyncssh.misc.KeyExchangeFailed: Major (720896): The referenced credential has expired, Minor (100001): Unknown code 0
```

I can reproduce the problem with a modified version of the simple client from asyncssh’s examples

```python
import asyncio, asyncssh, sys, logging

async def run_client() -> None:
    async with asyncssh.connect('host') as conn:
        while True:
            result = await conn.run('klist', check=True)
            print(result.stdout, end='')
            await asyncio.sleep (60)

logging.basicConfig (level=logging.DEBUG)
asyncssh.set_debug_level (2)
try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

with reduced ticket lifetime (2 minutes), automated renewal and reduced rekey time (`RekeyLimit 1M 60`). This confirms that re-keying is indeed the issue:

```
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0]   Key exchange algs: gss-group14-sha256-toWM5Slw5Ew8Mqkay+al2g==,gss-group16-sha512-toWM5Slw5Ew8Mqkay+al2g==,gss-nistp256-sha256-toWM5Slw5Ew8Mqkay+al2g==,gss-curve25519-sha256-toWM5Slw5Ew8Mqkay+al2g==,gss-group14-sha1-toWM5Slw5Ew8Mqkay+al2g==,gss-gex-sha1-toWM5Slw5Ew8Mqkay+al2g==,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,sntrup761x25519-sha512@openssh.com,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
DEBUG:asyncssh:[conn=0]   Host key algs: rsa-sha2-512,rsa-sha2-256,ecdsa-sha2-nistp256,ssh-ed25519
DEBUG:asyncssh:[conn=0]   Client to server:
DEBUG:asyncssh:[conn=0]     Encryption algs: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
DEBUG:asyncssh:[conn=0]     MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib@openssh.com
DEBUG:asyncssh:[conn=0]   Server to client:
DEBUG:asyncssh:[conn=0]     Encryption algs: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
DEBUG:asyncssh:[conn=0]     MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib@openssh.com
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0]   Key exchange alg: gss-curve25519-sha256
INFO:asyncssh:[conn=0] Sending disconnect: Major (720896): The referenced credential has expired, Minor (100001): Unknown code 0 (3)
INFO:asyncssh:[conn=0] Connection failure: Major (720896): The referenced credential has expired, Minor (100001): Unknown code 0
INFO:asyncssh:[conn=0, chan=2] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=2] Channel closed: Major (720896): The referenced credential has expired, Minor (100001): Unknown code 0
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
```

However I cannot reproduce the issue with plain SSH:

```bash
ssh -v $HOST -- 'while true; do klist; sleep 60; done'
```

So I looked at [gss_unix.py](https://github.com/ronf/asyncssh/blob/master/asyncssh/gss_unix.py) and while its `reset` method (which is called [here](https://github.com/ronf/asyncssh/blob/master/asyncssh/connection.py#L2123)) resets the `SecurityContext`, it does not set up fresh `Credentials`. I’m assuming it’s trying to re-key using existing, expired credentials and thus fails. Since gssapi does not have a method to reset itself, it looks like it might be necessary to create a new object like so:

```diff
diff --git a/asyncssh/gss_unix.py b/asyncssh/gss_unix.py
index 4d6c5a3..8401baa 100644
--- a/asyncssh/gss_unix.py
+++ b/asyncssh/gss_unix.py
@@ -37,19 +37,14 @@ def _mech_to_oid(mech):
 class _GSSBase:
     """GSS base class"""

-    def __init__(self, host, usage):
+    def __init__(self, host):
         if '@' in host:
             self._host = Name(host)
         else:
             self._host = Name('host@' + host, NameType.hostbased_service)

-        if usage == 'initiate':
-            self._creds = Credentials(usage=usage)
-        else:
-            self._creds = Credentials(name=self._host, usage=usage)
-
-        self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
         self._ctx = None
+        self._mechs = None

     def _init_context(self):
         """Abstract method to construct GSS security context"""
@@ -60,6 +55,9 @@ class _GSSBase:
     def mechs(self):
         """Return GSS mechanisms available for this host"""

+        if not self._mechs:
+            self._init_context()
+
         return self._mechs

     @property
@@ -97,6 +95,7 @@ class _GSSBase:
         """Reset GSS security context"""

         self._ctx = None
+        self._creds = None

     def step(self, token=None):
         """Perform next step in GSS security exchange"""
@@ -125,7 +124,7 @@ class GSSClient(_GSSBase):
     """GSS client"""

     def __init__(self, host, delegate_creds):
-        super().__init__(host, 'initiate')
+        super().__init__(host)

         flags = set((RequirementFlag.mutual_authentication,
                      RequirementFlag.integrity))
@@ -138,6 +137,8 @@ class GSSClient(_GSSBase):
     def _init_context(self):
         """Construct GSS client security context"""

+        self._creds = Credentials(usage='initiate')
+        self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
         self._ctx = SecurityContext(name=self._host, creds=self._creds,
                                     flags=self._flags)
@@ -146,9 +147,11 @@ class GSSServer(_GSSBase):
     """GSS server"""

     def __init__(self, host):
-        super().__init__(host, 'accept')
+        super().__init__(host)

     def _init_context(self):
         """Construct GSS server security context"""

+        self._creds = Credentials(name=self._host, usage='accept')
+        self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
         self._ctx = SecurityContext(creds=self._creds)
```

This fixes the disconnects, but causes alot of tests to fail, so we have to come up with a different solution.

Please let me know if you need further information or if I can help reproducing the issue.

**Comment by ronf** [comment]
Thanks for the report and the detailed proposed fix!

The changes make sense to me, and I'll look into why the unit tests are failing after this change. My guess is that the GSS stub code in the tests directory might need to be adjusted, though I'm not sure of the specific issue there without seeing the unit test failures. I'll try to take a closer look at this tonight.

Do you happen to have any Windows clients in your setup? I'm curious if the gss_win32 code has a  similar issue or not. It doesn't have a separate Credentials object, but it looks like it only creates the ClientAuth/ServerAuth objects at startup. So, it would depend on whether those objects were able to take care of refreshing the credentials automatically when authorize() is called from the step() function there.

**Comment by ronf** [comment]
I took a closer look at this tonight. The problem with the unit tests appears to be related to code querying the `mechs` property on the GSS objects. The new code was attempting to create a SecurityContext before returning the mechs, and that seems to be causing problems.

I was able to change the code to create an empty `Credentials` object that works for both GSSClient and GSSServer with no parameters, and use that to fill in the `mechs` when initializing the objects, and later create a fresh `Credentials` object with appropriate parameters in the `_init_context` method when a `SecurityContext` was needed and that seemed to fix all the unit test errors.

Here's an updated diff, meant to be applied against the latest version of gss_unix.py:

```patch
diff --git a/asyncssh/gss_unix.py b/asyncssh/gss_unix.py
index bc34bad..ac595f8 100644
--- a/asyncssh/gss_unix.py
+++ b/asyncssh/gss_unix.py
@@ -39,18 +39,13 @@ def _mech_to_oid(mech: OID) -> bytes:
 class GSSBase:
     """GSS base class"""

-    def __init__(self, host: str, usage: str):
+    def __init__(self, host: str):
         if '@' in host:
             self._host = Name(host)
         else:
             self._host = Name('host@' + host, NameType.hostbased_service)

-        if usage == 'initiate':
-            self._creds = Credentials(usage=usage)
-        else:
-            self._creds = Credentials(name=self._host, usage=usage)
-
-        self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
+        self._mechs = [_mech_to_oid(mech) for mech in Credentials().mechs]
         self._ctx: Optional[SecurityContext] = None

     def _init_context(self) -> None:
@@ -141,7 +136,7 @@ class GSSClient(GSSBase):
     """GSS client"""

     def __init__(self, host: str, delegate_creds: bool):
-        super().__init__(host, 'initiate')
+        super().__init__(host)

         flags = set((RequirementFlag.mutual_authentication,
                      RequirementFlag.integrity))
@@ -154,17 +149,18 @@ class GSSClient(GSSBase):
     def _init_context(self) -> None:
         """Construct GSS client security context"""

-        self._ctx = SecurityContext(name=self._host, creds=self._creds,
+        creds = Credentials(usage='initiate')
+
+        self._ctx = SecurityContext(name=self._host, creds=creds,
                                     flags=self._flags)


 class GSSServer(GSSBase):
     """GSS server"""

-    def __init__(self, host: str):
-        super().__init__(host, 'accept')
-
     def _init_context(self) -> None:
         """Construct GSS server security context"""

-        self._ctx = SecurityContext(creds=self._creds)
+        creds = Credentials(name=self._host, usage='accept')
+
+        self._ctx = SecurityContext(creds=creds)
```

Note that this is newer than the version your diff was against. It looks like you edited the version before type annotations were added.

Could you give this a try and let me know if it fixes the issue you were seeing?

Thanks again for your report!

**Comment by PromyLOPh** [comment]
No, your patch does not work for me unfortunately. When manually running `Credentials().mechs` I get this error

```
gssapi.raw.exceptions.MissingCredentialsError: Major (458752): No credentials were supplied, or the credentials were unavailable or inaccessible, Minor (2529639093): Keytab FILE:/etc/krb5.keytab is nonexistent or empty
```

and asyncssh does not even try GSSAPI authentication (and thus the login fails). I’m assuming we have to catch `MissingCredentialsError` when there are no credentials available. This one works for me and the tests still pass:

```diff
diff --git a/asyncssh/gss_unix.py b/asyncssh/gss_unix.py
index bc34bad..dfcc1c2 100644
--- a/asyncssh/gss_unix.py
+++ b/asyncssh/gss_unix.py
@@ -24,7 +24,7 @@ from typing import Optional, Sequence

 from gssapi import Credentials, Name, NameType, OID
 from gssapi import RequirementFlag, SecurityContext
-from gssapi.exceptions import GSSError
+from gssapi.exceptions import GSSError, MissingCredentialsError

 from .asn1 import OBJECT_IDENTIFIER

@@ -39,18 +39,13 @@ def _mech_to_oid(mech: OID) -> bytes:
 class GSSBase:
     """GSS base class"""

-    def __init__(self, host: str, usage: str):
+    def __init__(self, host: str):
         if '@' in host:
             self._host = Name(host)
         else:
             self._host = Name('host@' + host, NameType.hostbased_service)

-        if usage == 'initiate':
-            self._creds = Credentials(usage=usage)
-        else:
-            self._creds = Credentials(name=self._host, usage=usage)
-
-        self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
+        self._mechs = None
         self._ctx: Optional[SecurityContext] = None

     def _init_context(self) -> None:
@@ -62,6 +57,9 @@ class GSSBase:
     def mechs(self) -> Sequence[bytes]:
         """Return GSS mechanisms available for this host"""

+        if not self._mechs:
+            self._init_context ()
+
         return self._mechs

     @property
@@ -106,6 +104,7 @@ class GSSBase:
     def reset(self) -> None:
         """Reset GSS security context"""

+        self._mechs = None
         self._ctx = None

     def step(self, token: Optional[bytes] = None) -> Optional[bytes]:
@@ -141,7 +140,7 @@ class GSSClient(GSSBase):
     """GSS client"""

     def __init__(self, host: str, delegate_creds: bool):
-        super().__init__(host, 'initiate')
+        super().__init__(host)

         flags = set((RequirementFlag.mutual_authentication,
                      RequirementFlag.integrity))
@@ -154,17 +153,25 @@ class GSSClient(GSSBase):
     def _init_context(self) -> None:
         """Construct GSS client security context"""

-        self._ctx = SecurityContext(name=self._host, creds=self._creds,
-                                    flags=self._flags)
+        try:
+            creds = Credentials(usage='initiate')
+            self._mechs = [_mech_to_oid(mech) for mech in creds.mechs]
+            self._ctx = SecurityContext(name=self._host, creds=creds,
+                                        flags=self._flags)
+        except MissingCredentialsError:
+            # If the user has no valid GSSAPI credentials we cannot authenticate using GSSAPI.
+            self._mechs = []


 class GSSServer(GSSBase):
     """GSS server"""

-    def __init__(self, host: str):
-        super().__init__(host, 'accept')
-
     def _init_context(self) -> None:
         """Construct GSS server security context"""

-        self._ctx = SecurityContext(creds=self._creds)
+        try:
+            creds = Credentials(name=self._host, usage='accept')
+            self._mechs = [_mech_to_oid(mech) for mech in creds.mechs]
+            self._ctx = SecurityContext(creds=creds)
+        except MissingCredentialsError:
+            self._mechs = []
```

I don’t have a Windows machine available and don’t know how to operate Kerberos on Windows either. Sorry about the diff, I was using v2.7.1, because that’s packaged in GNU Guix. I switched to v2.10.1 now.

**Comment by ronf** [comment]
Thanks for doing the test.

Something is not adding up here -- the original code didn't need to explicitly catch the MissingCredentialsError, so why should the new version be different in that regard?

I have an alternate fix here which passes all the unit tests, but when I manually test it against the _real_ GSSAPI library, I'm running into issues getting it to work in some cases. I'm seeing the same errors with the prior gss_unix.py which used to work, though, so I think this may be a consequence of switching machines recently. I may not have gotten the Kerberos environment set up quite right on the new systems.

If you'd like to try another version in the mean time, here's my latest:

```patch
diff --git a/asyncssh/gss_unix.py b/asyncssh/gss_unix.py
index bc34bad..bc1d49f 100644
--- a/asyncssh/gss_unix.py
+++ b/asyncssh/gss_unix.py
@@ -39,20 +39,21 @@ def _mech_to_oid(mech: OID) -> bytes:
 class GSSBase:
     """GSS base class"""

-    def __init__(self, host: str, usage: str):
+    def __init__(self, host: str):
         if '@' in host:
             self._host = Name(host)
         else:
             self._host = Name('host@' + host, NameType.hostbased_service)

-        if usage == 'initiate':
-            self._creds = Credentials(usage=usage)
-        else:
-            self._creds = Credentials(name=self._host, usage=usage)
-
         self._mechs = [_mech_to_oid(mech) for mech in self._creds.mechs]
         self._ctx: Optional[SecurityContext] = None

+    @property
+    def _creds(self) -> None:
+        """Abstract method to construct GSS credentials"""
+
+        raise NotImplementedError
+
     def _init_context(self) -> None:
         """Abstract method to construct GSS security context"""

@@ -141,7 +142,7 @@ class GSSClient(GSSBase):
     """GSS client"""

     def __init__(self, host: str, delegate_creds: bool):
-        super().__init__(host, 'initiate')
+        super().__init__(host)

         flags = set((RequirementFlag.mutual_authentication,
                      RequirementFlag.integrity))
@@ -151,6 +152,12 @@ class GSSClient(GSSBase):

         self._flags = flags

+    @property
+    def _creds(self) -> Credentials:
+        """Abstract method to construct GSS credentials"""
+
+        return Credentials(usage='initiate')
+
     def _init_context(self) -> None:
         """Construct GSS client security context"""

@@ -161,8 +168,11 @@ class GSSClient(GSSBase):
 class GSSServer(GSSBase):
     """GSS server"""

-    def __init__(self, host: str):
-        super().__init__(host, 'accept')
+    @property
+    def _creds(self) -> Credentials:
+        """Abstract method to construct GSS credentials"""
+
+        return Credentials(name=self._host, usage='accept')

     def _init_context(self) -> None:
         """Construct GSS server security context"""
```

I'll try and fix my setup here and test it using the original code and then try the patch again once that's working.

**Comment by PromyLOPh** [comment]
This patch indeed works and fixes the initial problem for me. If I set up invalid Kerberos credentials like `export KRB5CCNAME=FILE:/nonexistent` then manually instantiating `Credentials(usage='initiate')` errors out with `gssapi.raw.exceptions.MissingCredentialsError` while asyncssh just ignores GSSAPI completely. Maybe that exception is caught somewhere (implicitly?) and with my 2nd patch the instantiation was moved to a code branch not protected by an try-except clause.

**Comment by ronf** [comment]
Thanks for confirming the patch. I'll do a test this weekend with an invalid credential cache via KRB5CCNAME and see if I can reproduce the issue you are seeing with needing to catch MissingCredentialsError. I don't see this error on macOS even with a bad KRB5CCNAME, but perhaps that's been a problem on Linux all along that I just never noticed.

**Comment by ronf** [comment]
Ok - I've pushed this patch and a fix for another bug I found during my testing to the "develop" branch. See commits 815eca0 and cbe258a. Thanks again for reporting this issue and suggesting the fix! Let me know if you run into any issues with this.

I tried reproducing the MissingCredentialsError issue on both Linux and macOS, but setting an invalid KRB5CCNAME doesn't seem to trigger any error for me. It just falls back to non-Kerberos auth if allowed, or returns "Permission denied" if I request that it only use Kerberos auth. Let me know if you're still seeing an issue with this. Perhaps there's some other difference in our environments which could explain it.

**Comment by PromyLOPh** [comment]
Thanks for pushing the fix, the develop branch works and fixes the issue.

I think the MissingCredentialsError issue might be a misunderstanding. asyncssh indeed just moves on to non-GSSAPI-based authentication. Manually creating an instance of `Credentials` produces said error. But that is not relevant here, because asyncssh does the right thing.

**Comment by ronf** [comment]
Great - thanks again for reporting this, and helping with the testing! This will be included in the next AsyncSSH release.

**Comment by ronf** [comment]
This change is now available in the AsyncSSH 2.11.0 release.
